bindings/rust/src/lazy_reader.rs

```diff
@@ -0,0 +1,518 @@
+            return Err(Error::Verify(format!(
+                "verify_fixvec: self.size({}) < NUMBER_SIZE(4)",
+                self.size
+            )));
+        }
+        let item_count = self.unpack_number()?;
+        if item_count == 0 {
+            if self.size == NUMBER_SIZE {
+                return Ok(());
+            } else {
+                return Err(Error::Verify(format!(
+                    "verify_fixvec: self.size({}) != 4",
+                    self.size
+                )));
+            }
+        }
+
+        let total_size = calculate_offset(item_size, item_count, NUMBER_SIZE)?;
+        if self.size == total_size {
+            Ok(())
+        } else {
+            Err(Error::Verify(format!(
+                "verify_fixvec: self.size({}) != total_size({})",
+                self.size, total_size
+            )))
+        }
+    }
+
+    pub fn option_is_none(&self) -> bool {
+        self.size == 0
+    }
+    pub fn fixvec_length(&self) -> Result<usize, Error> {
+        self.unpack_number()
+    }
+
+    pub fn dynvec_length(&self) -> Result<usize, Error> {
+        if self.size == NUMBER_SIZE {
+            Ok(0)
+        } else {
+            let mut cur2 = self.clone();
+            cur2.add_offset(NUMBER_SIZE)?;
+            cur2.sub_size(NUMBER_SIZE)?;
+            cur2.validate()?;
+            cur2.get_item_count()
+        }
+    }
+
+    pub fn get_item_count(&self) -> Result<usize, Error> {
+        let len = self.unpack_number()?;
+        if len % 4 != 0 {
+            return Err(Error::UnknownItem(format!(
+                "get_item_count: not aligned, len = {}",
+                len
+            )));
+        }
+        let count = len / 4;
+        if count == 0 {
+            Err(Error::UnknownItem("get_item_count".into()))
+        } else {
+            Ok(count - 1)
+        }
+    }
+
+    pub fn table_actual_field_count(&self) -> Result<usize, Error> {
+        self.dynvec_length()
+    }
+
+    pub fn table_has_extra_fields(&self, field_count: usize) -> Result<bool, Error> {
+        let count = self.table_actual_field_count()?;
+        Ok(count > field_count)
+    }
+
+    pub fn slice_by_offset(&self, offset: usize, size: usize) -> Result<Cursor, Error> {
+        let mut cur2 = self.clone();
+        cur2.add_offset(offset)?;
+        cur2.size = size;
+        cur2.validate()?;
+        Ok(cur2)
+    }
+
+    pub fn fixvec_slice_by_index(
+        &self,
+        item_size: usize,
+        item_index: usize,
+    ) -> Result<Cursor, Error> {
+        let mut cur2 = self.clone();
+        let item_count = self.unpack_number()?;
+        if item_index >= item_count {
+            Err(Error::OutOfBound(format!(
+                "fixvec_slice_by_index: item_index({}) >= item_count({})",
+                item_index, item_count
+            )))
+        } else {
+            let offset = calculate_offset(item_size, item_index, NUMBER_SIZE)?;
+            cur2.add_offset(offset)?;
```

Comment:
I see a lot of patterns like this, where we clone a cursor, add something to its offset, then manipulate the size by directly changing it, or via sub_size. Are we really doing slicing here? Can we fix those code by a single slicing operation(and also removing both `add_offset` and `sub_size`)? That will simplify the code a lot and reduce the attack vector

Reply:
Sometime we need to adjust cursor manually, see c code: https://github.com/XuJiandong/omnilock/blob/90ad8ba6744b9c5d261c713b6b71c30740d54a98/c/cobuild.c#L269-L271

Reply:
To me a slicing operation is all you need here. Depending on the exact API of slicing operation, we can do one of the following 2:

```
mol2_slice_by_length(cursor, offset, cursor->length() - offset);
```

Or

```
mol2_slice_by_end(cursor, offset, cursor->end());
```

Reply:
If you check "calculate_offset" carefully: they shows in code 4 times and shares nothing common.
We can't get any benefit doing so. note: there are `slice_by_offset`, `slice_by_index` already

Reply:
I don't agre with this assessment. We are mixing 2 independent things together:

* How do I safely get a slice of the data from a cursor, given an offset and either a length, or and end value?
* Molecule uses indices to maintain molecule offsets, how do I calculate the offsets of a certain field?

The fact that we have `slice_by_offset`, `slice_by_index` and other functions all working on a single cursor type feels extremely messy to me. If I were doing this, I would design 2 separate components:

* A cursor only has a slicing operation that takes an offset and a length, internally, it checks if the offset and length are vaild, if not, an error would be thrown. If they are indeed valid, a new cursor will be returned.
* An offset calculation function that only calculates the offset for dynvec or a table. If you really want it, you can have `slice_by_index` function that first calculates the offset, then calls the slicing operation above to build the final cursor. I personally find it messy and possible dangerous to have a `dynvec_slice_by_index` function, that directly manipulates a cursor, this is not a proper abstraction. When the cursor logic requires altering, we will need to touch a ton of functions here, ensuring they all obey cursor's validation rules. This is a nightmare to maintain.